### PR TITLE
NAC-432: Display bundles list in Admin Console

### PR DESCRIPTION
**Pull request by naveen-konda**
Jira: [NAC-432](https://hyland.atlassian.net/browse/NAC-432)

## Summary

Adds a **Bundles** page so administrators can inspect the bundles deployed on a running instance from the Admin Console instead of falling back to the legacy JSF UI.

The page reads `GET /management/distribution` and renders:

- a **Distribution** card (application name/version, distribution name/version/date)
- a **Bundles** table (name, version, revision) with pagination and a client-side filter
- **Warnings** and **Errors** cards, each shown only when the server reports any

Loading, error and no-data states are all handled; the error state offers a retry and raises the shared server-error modal.

## Notes for reviewers

- **No store slice.** The data is read-only and used by one page, so this follows the `configuration-properties` pattern of a plain service call with `isDataLoaded` / `isError` flags rather than adding NgRx actions/reducers/effects.
- **Response shape** was taken from the platform's `SimplifiedServerInfoJsonWriter` rather than inferred from a sample payload. That is where the optional `revision` field comes from: the server only writes it for bundles built from a source revision (in practice, Studio packages), so the column falls back to `N/A`.
- **`distributionDate` is displayed verbatim.** The platform returns it as a free-form framework property that can even be the literal string `unknown`, so parsing it risked rendering an incorrect date. The legacy JSF page also showed it raw.
- **Filter is bound to `(input)`, not `(keyup)`**, so it also reacts to paste, autofill and clearing. A DOM-level test dispatches the real event to pin this down.

## Test plan

- [x] `npm run lint` passes
- [x] `npm test` passes: 875 tests across 66 files (16 new, covering the service endpoint, the load/error/no-data paths, the distribution summary fallbacks, filtering and paginator wiring)
- [x] `npm run build` succeeds, with `bundles-module` emitted as its own lazy chunk
- [x] Verified manually against a live Nuxeo instance reporting 128 bundles:
  - [x] Distribution card, table, paginator (`1 – 25 of 128`) and `Warnings (7)` render; no Errors card when the server reports zero
  - [x] Filtering `ecm.core` narrows to 22 rows and resets the paginator to page 1
  - [x] A non-matching term shows "No bundles match the applied filter." with a `0 of 0` paginator
  - [x] Clearing the filter restores all 128 bundles
  - [x] Revision column populates for Studio bundles
  - [x] `Bundles` appears in the left navigation and no console errors occur

Made with [Cursor](https://cursor.com)

### CODE REVIEW

**Comment by narasimhahyland** [comment]
### 1. A response with no bundles hides the Distribution, Warnings and Errors cards — `should-fix`

`features/bundles/bundles.component.html` lines 30–128

The Warnings and Errors blocks are nested inside the block gated on `hasBundles()`:

```html
@if (isDataLoaded && !isError && hasBundles()) {
  <mat-card>Distribution</mat-card>
  <mat-card>Bundles table</mat-card>
  @if (warnings.length > 0) { … }
  @if (errors.length > 0) { … }
}
```

`getBundles()` populates `warnings`/`errors` independently of `bundles`, so when the server returns an empty or absent `bundles` array alongside real contribution errors, the page renders only "Bundle details are not available." and drops the errors, the warnings, and the distribution identity — the information an admin came to this page for. The component already treats a bundle-less payload as reachable (`data?.bundles || []`, plus a dedicated test feeding `of({})`), but that test asserts `hasBundles() === false` and never checks the template, which is why CI is green.

**Fix:** gate the outer block on "a response arrived" rather than on the bundle count, and wrap only the table card in `@if (hasBundles())`, with the no-data message inside it. Add a test with `bundles: []` and non-empty `warnings`/`errors`.

### 2. `DistributionResponse` declares required fields the code treats as optional, forcing casts that disable type checking — `should-fix`

`shared/types/bundles.interface.ts`

Every field except `entity-type` and `revision` is declared required, while every consumer optional-chains with a fallback: `data?.bundles || []`, `data?.warnings || []`, `field.value || NOT_AVAILABLE`, and `bundle?.version || N/A` for a `version` the interface declares required. The spec can only construct a partial payload by writing `{} as DistributionResponse` — twice — which switches off the compiler exactly where a real server sending a partial response is plausible. `revision?: string` is the one field where declaration and handling agree.

**Fix:** mark the defensively-handled fields optional (`applicationName?`, `applicationVersion?`, `distributionName?`, `distributionVersion?`, `bundles?`, `warnings?`, `errors?`, `BundleInfo.version?`). The `as DistributionResponse` casts then become unnecessary and the compiler verifies the fallbacks.

### 3. No test asserts any template state, so the rendering logic is unverified — `should-fix`

`features/bundles/bundles.component.spec.ts`

Only `should create` and `should load bundles and distribution details on init` call `fixture.detectChanges()`, both on the success path. Every state test asserts flags only (`expect(component.isError).toBe(true)`). None of the four `@if` conditions is exercised: you could invert the loading and error guards, or delete the `hasBundles()` guard behind issue 1, and the suite would stay green. The retry button — a headline behaviour in the PR description — is never clicked; all error tests call `component.getBundles()` directly.

**Fix:** add DOM assertions for `.loading-container`, `.error-container` and `.no-data-container`, and a test that clicks Retry and asserts the service is called again and `isError` clears.

### 4. Three tests pass without verifying the behaviour they name — `minor`

`features/bundles/bundles.component.spec.ts` and `services/bundles.service.spec.ts`

- `should complete destroy$ subject on ngOnDestroy` spies on `destroy$.next`/`complete` then calls `ngOnDestroy()`, which does exactly those two things. It still passes with `takeUntil(this.destroy$)` deleted from `getBundles()`, so the unsubscription it exists to prove is untested.
- `should filter bundles by name, version and revision` would pass unchanged under Material's default predicate: the mock bundles' only own keys are `name`, `version` and `revision`, so the custom `filterPredicate` and the default concatenation behave identically for all three assertions. Deleting the constructor leaves CI green. The predicate's one real behavioural difference — ignoring bundle fields the server sends but the table doesn't display — is the case not covered.
- `should call makeHttpRequest with the correct endpoint` asserts only that the string `"GET_DISTRIBUTION_INFO"` was forwarded, not that it resolves to `GET /management/distribution`. A wrong URL would pass, as would a missing `method` — which `NetworkService` silently turns into a `PUT` via `const method = config.method || "PUT"`. The config is correct today, so this is a coverage gap rather than a live bug. The mock also returns `undefined` and the test never asserts the observable is passed through.

**Fix:** assert `REST_END_POINT_CONFIG.GET_DISTRIBUTION_INFO` equals `{ endpoint: "/management/distribution", method: "GET" }` and that the returned observable is forwarded; add a bundle with a non-displayed property to the filter test; assert teardown by emitting through the mocked service after `ngOnDestroy()`.

### 5. `distributionDate` is dead, and the PR body still advertises it — `minor`

`shared/types/bundles.interface.ts:18`

The fourth commit ("remove distribution date field") dropped the date from the summary, but `distributionDate: string` remains declared required on the interface and set in the spec mock, and is read nowhere at head. The PR description still says the Distribution card shows "distribution name/version/**date**", and the Jira note explaining the date is "displayed verbatim" is now stale. Because it is only a type annotation, neither `tsc` nor ESLint flags it.

**Fix:** drop the field from the interface and the mock, and correct the PR body and ticket note — or restore it to the summary if the removal was accidental.

### 6. The bundle count in the header ignores the filter — `minor`

`features/bundles/bundles.component.html:47`

```html
{{ buildLabelWithCount(BUNDLES_LABELS.BUNDLES_COUNT, bundlesData.data.length) }}
```

The header binds the unfiltered `data.length` while the paginator on line 97 binds `filteredData.length`. Filtering 128 bundles to 3 shows "Bundles (128)" directly above a paginator reading "1 – 3 of 3", and the "No bundles match the applied filter." message can appear under a large count.

**Fix:** bind `filteredData.length`, or show both as "3 of 128" when a filter is active.

### 7. `sticky` on the header cell does nothing — `minor`

`features/bundles/bundles.component.html:59`

```html
<th mat-header-cell *matHeaderCellDef sticky>
```

`sticky` is an input of `CdkColumnDef` (selector `[cdkColumnDef]`, i.e. the `<ng-container matColumnDef="name">`), not of `CdkHeaderCell` or `CdkHeaderCellDef` — confirmed against the `@angular/cdk@21.0.2` typings. As a bare attribute on the `th` it binds to nothing, so the Name column is not pinned when the 550px-max-height, `overflow: auto` container scrolls horizontally. The sticky header *row* works, but that comes entirely from `*matHeaderRowDef="columnsToDisplay; sticky: true"`. The same inert attribute exists at `probes-data.component.html:18`, so this is inherited rather than new.

**Fix:** move `sticky` to `<ng-container matColumnDef="name" sticky>`, or delete it if column pinning isn't wanted.

### 8. The paginator setter never clears a destroyed paginator — `minor`

`features/bundles/bundles.component.ts:38`

```ts
@ViewChild(MatPaginator) set tablePaginator(paginator: MatPaginator) {
  if (paginator) { this.bundlesData.paginator = paginator; }
}
```

The guard exists because Angular calls the setter with `undefined` when the `@if` tears the paginator down, but the missing else branch leaves `bundlesData.paginator` referencing a destroyed instance until a new one mounts. Not reachable today — the only re-fetch is Retry, which only exists in the error state where the table was never rendered — but adding a Refresh button or a poll to the loaded state would make it live.

**Fix:** assign unconditionally: `this.bundlesData.paginator = paginator ?? null;`

### 9. Accessibility details — `minor`

`features/bundles/bundles.component.html`

- `tabindex="0"` on all four `<mat-card>` regions (lines 31, 43, 104, 117) and on every `<tr mat-row>` (line 74) makes them focusable with no click or key handler, adding up to 100+ inert tab stops on a full page. The `tr:focus` background rule is the only thing it drives.
- `aria-label="Bundles Table"`, `"Select page"`, `"Bundle warnings"` and `"Bundle errors"` are inline literals while every other user-facing string comes from `BUNDLES_LABELS`.
- `aria-live="polite"` sits on the static `<h1>`, which never changes, while the loading, error and no-data containers that actually replace one another carry no live region. Only `.bundles__no-match` has one.

The `tabindex` and `aria-live` patterns are copied from `probes-data` and `probes`, so they're house style rather than new deviations.

**Fix:** drop `tabindex="0"` from the non-interactive cards and rows, move the aria strings into `BUNDLES_LABELS`, and put the live region on the state container rather than the title.

**Comment by naveen-konda** [comment]
Thanks for the thorough review @narasimhahyland — every point held up, so I've addressed all nine in cfc5a89. Details below, numbered to match your comment.

### 1. Empty `bundles` hides Distribution/Warnings/Errors — fixed
The outer block is now gated on `isDataLoaded && !isError` (a response arrived) rather than on the bundle count. The Distribution card, Warnings and Errors now always render for a successful response, and only the bundles table is wrapped in `@if (hasBundles())`, with the no-data message shown in the `@else` inside the table card. Added a test (`should show the no-data message inside the card when the response has no bundles`) that feeds `bundles: []` with non-empty `warnings`/`errors` and asserts the table is gone while the summary, warnings and errors cards all render.

### 2. `DistributionResponse` required fields vs. optional handling — fixed
Marked the defensively-handled fields optional (`applicationName?`, `applicationVersion?`, `distributionName?`, `distributionVersion?`, `bundles?`, `warnings?`, `errors?`, and `BundleInfo.version?`). The two `{} as DistributionResponse` casts are gone — the spec now uses plain `{}` and the compiler verifies the `|| fallback` paths.

### 3. No template-state assertions — fixed
Added DOM-level tests: `.loading-container` before a response arrives, `.error-container` on failure with a real **Retry** button `.click()` that asserts the service is re-called and `isError` clears, and `.no-data-container` for the bundle-less case. So inverting/deleting the state guards now fails the suite.

### 4. Three tests that didn't verify their behaviour — fixed
- **Teardown:** added `should stop applying distribution responses after ngOnDestroy`, which emits through the mocked service *after* `ngOnDestroy()` and asserts the table stays empty — this fails if `takeUntil(this.destroy$)` is removed.
- **Filter predicate:** added a bundle with a non-displayed `description` field and asserted a filter on that value matches nothing, which fails under Material's default (concatenate-all) predicate.
- **Service:** now asserts `REST_END_POINT_CONFIG.GET_DISTRIBUTION_INFO` equals `{ endpoint: "/management/distribution", method: "GET" }` and that the observable from `NetworkService` is forwarded verbatim.

### 5. Dead `distributionDate` — fixed
Removed from the interface and the spec mock (grep confirms no remaining references). I'll update the PR description/Jira note to drop the mention of "date" as well.

### 6. Header count ignored the filter — fixed
The header now binds `bundlesData.filteredData.length`, matching the paginator, so a 128→3 filter reads "Bundles (3)".

### 7. Inert `sticky` on the `<th>` — fixed
Moved `sticky` onto `<ng-container matColumnDef="name" sticky>` so the Name column actually pins on horizontal scroll (the header row keeps its `sticky: true`).

### 8. Paginator setter never cleared — fixed
The setter now assigns unconditionally (`set tablePaginator(paginator: MatPaginator | null)` → `this.bundlesData.paginator = paginator`), so a torn-down paginator is cleared rather than left dangling for a future refresh/poll.

### 9. Accessibility — fixed
Dropped the inert `tabindex="0"` from the cards and `<tr mat-row>`, moved `aria-label` strings ("Bundles table", "Select page", "Bundle warnings", "Bundle errors") into `BUNDLES_LABELS.ARIA_LABELS`, and moved `aria-live="polite"` off the static `<h1>` onto the loading/error state containers that actually swap.

Verification: `eslint` clean on the touched files, `vitest` green (23/23 for the bundles specs), and `ng build` succeeds. PTAL.

File: nuxeo-admin-console-web/angular-app/src/app/features/bundles/bundles.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,365 @@
...
+
+  it("should always render the warnings and errors cards with a placeholder when empty", () => {
+    vi.spyOn(bundlesService, "getDistributionInfo").mockReturnValue(
+      of({ bundles: mockDistribution.bundles, warnings: [], errors: [] })
+    );
+    fixture.detectChanges();
+    const host = fixture.nativeElement;
+    const emptyMessages = host.querySelectorAll(".bundles__empty-messages");
+    expect(emptyMessages.length).toBe(2);
+    expect(emptyMessages[0].textContent).toContain(
```
Comment (madhurkulshrestha-hyland) [inline nuxeo-admin-console-web/angular-app/src/app/features/bundles/bundles.component.spec.ts:212]: Rectify sonar issue

Comment (naveen-konda) [reply nuxeo-admin-console-web/angular-app/src/app/features/bundles/bundles.component.spec.ts:212]: Done, fixed.

**Comment by narasimhahyland** [comment]
Verified all nine against the code at `95171fc` rather than the reply, and they're genuinely fixed: the gating is hoisted out of `hasBundles()` with only the table wrapped, the interface fields are optional and both `{} as DistributionResponse` casts are gone, the three state branches have DOM assertions including a real Retry `.click()`, the teardown test fails if `takeUntil` is removed, the filter test fails under Material's default predicate, `distributionDate` is gone everywhere, the header binds `filteredData.length`, `sticky` sits on `<ng-container matColumnDef="name">`, the paginator setter is unconditional, and no `tabindex` or inline `aria-label` remains in the feature. CI is green.

Three things left, only the first of which I'd hold the merge for.

### 1. The PR description contradicts the code in three places — `should-fix`

The body was last updated at 06:32, before the final commits, and still says:

- "**Warnings** and **Errors** cards, each shown only when the server reports any" — as of `ce8c108` both cards always render with "No warnings reported." / "No errors reported." placeholders. The manual-verification checkbox "no Errors card when the server reports zero" describes behaviour that no longer exists.
- "**`distributionDate` is displayed verbatim.**" — the field is deleted from the interface and the mock and read nowhere. This bullet and the matching Jira note describe a field the PR no longer has.
- "16 new tests" / "875 tests across 66 files" — the two new bundles specs now hold 24 `it()` blocks (20 component, 4 service).

Worth correcting because the description is what lands in the merge record and what the next reader trusts.

### 2. Always-rendering Warnings/Errors cards — a decision to confirm, not a defect

My finding was that the cards must not be *hidden* when `bundles` is empty. `ce8c108` went further and made them unconditional, so a healthy instance permanently shows two cards whose entire content is "No warnings reported." and "No errors reported.", and a bundle-less response renders four cards, three of which say nothing happened. The conditional rendering was already correct once the outer block was gated on "a response arrived" instead of on the bundle count.

No code change needed if the placeholders are deliberate — just say so in the description and the ticket, since it reverses a documented decision. Otherwise restore `@if (warnings.length > 0)` inside the now-correct outer block.

### 3. The relocated `aria-live` regions still won't be announced — `minor`

`bundles.component.html:6,14`

```html
@if (!isDataLoaded && !isError) {
<div class="loading-container" aria-live="polite">
```

Moving the live region off the static `<h1>` was right, but the region is now created by the `@if` at the same moment as its content, and screen readers only announce mutations to a region already in the accessibility tree — one inserted together with its text is typically silent in NVDA and JAWS. My original suggestion said "put it on the state container" without noting that the container is itself conditional, so that half of the fix was incomplete advice on my part. The pre-existing `.bundles__no-match` on line 86 has the same shape.

Cheap to fix if you want it in this PR: keep one always-present `<div class="bundles__status" aria-live="polite">` and let the `@if` branches swap content inside it. Equally fine as a follow-up, since it matches the behaviour of the sibling pages today.

**Comment by naveen-konda** [comment]
Thanks @narasimhahyland — responses on the three follow-ups:

**1. PR description vs. code (should-fix):** Fixed — I've synced the PR description with the final code (the always-rendered warnings/errors placeholders, the removed `distributionDate`, and the corrected test count: 883 tests across 66 files, with the two new bundles specs adding 24). That said, these are description/summary edits with no bearing on code behaviour, so I'd treat this class of item as routine housekeeping rather than a review blocker.

**2. Always-rendering Warnings/Errors cards:** This is intentional. I'm deliberately showing the "No warnings reported." / "No errors reported." placeholders so an admin gets explicit, positive confirmation that the bundles/deployment are clean, rather than being left unsure whether the section simply failed to load. I've noted this decision in the PR description.

**3. `aria-live` regions:** Not a blocker / not required for this PR. The current pattern matches the existing `probes` / `probes-data` pages and isn't a regression (the previous `aria-live` sat on the static `<h1>`, which never announced state changes either). Fixing it only here would make the bundles page inconsistent with its siblings, so it's better handled as a shared a11y follow-up across those pages rather than in this PR.
